### PR TITLE
feat: SWOPP3 runner, plotting and deployment

### LINKED ISSUE #26: Implement the 8 SWOPP3 cases
## Context
SWOPP3 requires 8 cases per departure (Table 2 in the info package):

| Case | Name | Route | Strategy | WPS |
|------|------|-------|----------|-----|
| 1 | AO_WPS | Atlantic westbound | Optimised route + speed | with |
| 2 | AO_noWPS | Atlantic westbound | Optimised route + speed | without |
| 3 | AGC_WPS | Atlantic westbound | Great circle, fixed speed | with |
| 4 | AGC_noWPS | Atlantic westbound | Great circle, fixed speed | without |
| 5 | PO_WPS | Pacific eastbound | Optimised route + speed | with |
| 6 | PO_noWPS | Pacific eastbound | Optimised route + speed | without |
| 7 | PGC_WPS | Pacific eastbound | Great circle, fixed speed | with |
| 8 | PGC_noWPS | Pacific eastbound | Great circle, fixed speed | without |

## Requirements
- **Optimised (O)**: Run CMA-ES + FMS to find minimum-energy route and speed profile
- **Great Circle (GC)**: Fixed geodesic route, constant speed to meet passage time
- **WPS**: Use RISE polar model with wingsails enabled
- **noWPS**: Use RISE polar model with wingsails disabled (engine only)
- Each case has 366 departures → 8 × 366 = 2,928 total runs

## Dependencies
- #23 (RISE performance model)
- #24 (Route/departure configuration)
- #25 (ERA5 data)

## Acceptance Criteria
- [ ] Runner script that iterates over all 8 cases × 366 departures
- [ ] GC baseline implementation (fixed route, constant speed)
- [ ] Optimised route implementation with energy objective
- [ ] WPS/noWPS toggle via RISE polar API
- [ ] Parallel/batch execution support for large-scale runs

### PR DESCRIPTION
**Pull request by fjsuarez**
## Summary
Add SWOPP3 execution pipeline: runner, CLI, plotting, and SLURM deployment.

### New files
- `routetools/swopp3_runner.py`: Orchestrates optimization across 8 cases × 366 departures. Handles ERA5 loading, land mask creation, cost function assembly, CMA-ES optimization, and CSV output.
- `tests/test_swopp3_runner.py`: Runner integration tests
- `scripts/swopp3_run.py`: CLI entry point for local/HPC execution
- `scripts/swopp3_plot_routes.py`: Generate route comparison figures (optimized vs great-circle)
- `scripts/swopp3_slurm.sh`: SLURM submission script (CPU)
- `scripts/swopp3_slurm_gpu.sh`: SLURM submission script (GPU)

### Dependencies
All previous PRs (#41–#47)

---
*Part 8 of 9 — extracted from vangelis branch for clean review*

---
Closes #26, #30

### CODE REVIEW

File: routetools/swopp3_runner.py [hunk 1]
```diff
@@ -0,0 +1,593 @@
...
+    dlon = np.diff(lon)
+    dlat = np.diff(lat)
+    lat_mid = np.radians((lat[:-1] + lat[1:]) / 2)
+
+    # Approximate bearing using Mercator projection
+    dx = dlon * np.cos(lat_mid)  # eastward component
+    dy = dlat  # northward component
+
+    bearing = np.degrees(np.arctan2(dx, dy)) % 360.0
+    return bearing
```
Comment (daniprec) [inline routetools/swopp3_runner.py]: Does this not exist somewhere else in the library? I feel this should go in a more global module.

**Comment by daniprec** [comment]
Addressed review comments summary (code now pushed in `0041702`):

- `scripts/swopp3_run.py`
  - UTC-safe dataset epoch conversion is used (`datetime.fromtimestamp(..., tz=UTC).replace(tzinfo=None)`).
  - `xarray` dataset reads use context managers.
  - Explicit `KeyError` handling for missing lon/lat coordinate names.
  - Removed/avoided unused imports (`sys`).

- `scripts/swopp3_slurm.sh` / `scripts/swopp3_slurm_gpu.sh`
  - SBATCH logs now write to `slurm_%j.out/.err` (no pre-created output dir requirement).
  - CPU script preserves existing `XLA_FLAGS` while appending flags.

- `routetools/swopp3_runner.py`
  - Output directories are created before writing File A/B.
  - `evaluate_energy` now guards `curve.shape[0] >= 2` with clear `ValueError`.
  - Energy reduction keeps JAX type (`jnp.asarray` + `jnp.sum`) before scalar conversion.
  - `run_optimised_departure` now defaults `windfield=vectorfield` when needed for RISE cost.
  - Public helper used for bearings (`segment_bearings_deg`) with tests aligned.

- `scripts/swopp3_plot_routes.py`
  - Removed unused imports and dead helper concerns.
  - Added `--submission` support and auto-detection fallback for `IEUniversity-<n>-<case>.csv`.

- `tests/test_swopp3_runner.py`
  - Added regression tests for short-curve validation and vectorfield→windfield fallback behavior.

Proposed follow-up for “more general” functions (design comments):

1. Move route geometry helpers to a shared module (recommended: `routetools/route_geometry.py`):
   - `segment_bearings_deg`
   - any reusable segment midpoint/time helpers used across cost/runner code

2. Move generic energy-evaluation orchestration to `routetools/energy.py`:
   - a model-agnostic `evaluate_route_energy(...)` that takes sampled metocean + model callable
   - keep SWOPP3-specific wrappers in `swopp3_runner.py`

3. Move ERA5 dataset timestamp extraction to a shared helper in `routetools/era5/loader.py`:
   - `dataset_epoch(path: Path) -> datetime`
   - reused by runner and any other scripts that need robust epoch alignment

If this plan sounds good, I can open a small follow-up PR that performs only this extraction/refactor (no behavior changes).

**Comment by fjsuarez** [comment]
Thanks for the thorough fixes in `0041702` — all the Copilot review points have been addressed cleanly.

Regarding the proposed follow-up refactors:

1. **Route geometry helpers → `routetools/route_geometry.py`** — 👍 makes sense, `segment_bearings_deg` and midpoint helpers are reused across cost/runner/weather.
2. **Generic energy evaluation → `routetools/energy.py`** — 👍 good separation. Keep SWOPP3-specific wiring in `swopp3_runner.py`.
3. **ERA5 epoch helper → `routetools/era5/loader.py`** — 👍 agreed, `dataset_epoch(path)` belongs there.

Go ahead and open the follow-up PR when ready. LGTM on the current state — approving.

**Review by fjsuarez — COMMENTED** [review]
All Copilot review comments addressed in `0041702`. Follow-up refactor plan looks good — open a separate PR for the extraction. Ready to merge from my side.
